[iOS] Admin Dashboard - Device Management

JPKribs opened the pull request:
### Summary
Creates a DeviceView that allows for viewing all hardware that has connected to the Jellyfin Server and allows for assigning a custom name to them or deleting them.

On both the ViewModel and on the View, I do not allow the user to delete their own session. I think if were allowed to happen, then you would get 401'd and need to remove and re-add the user to get back in. 'Delete All' skips your device session. Not only does this mirror other clients with this functionality but it also avoids some weird interactions.

This should bring Swiftfin up to feature parity with Jellyfin-Web for the All Devices page features.

### Ideas
I think this is where I want this. The only feature I was realizing would be valuable is instead of a Delete All function, making the list so you can multselect, with a select all function on the list. Then, use that selection as the delete population. API-wise, there is no "Delete All" endpoint. It's always just been a forLoop to delete all Devices. There is no reason we couldn't make this a list pick instead. I stuck with Delete All to keep things matching but I'd be happy to look at the list idea too.

### Weird Behavior
Custom names on this SDK version are a little weird since we can **update** the customName but we cannot **see** the custom name. I have a TODO to update the deviceName to use CustomName when available. For now, you can update the custom name but it doesn't really take affect on the Swiftfin side so visually it's a little weird but it's all working fine behind the scene.

### Screenshots

<details>
  <summary>DevicesView</summary>

  <img src="https://github.com/user-attachments/assets/8ea41989-1374-4d65-8053-2d1be7a3ce1f" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-15 at 23:22:58" />
</details>

<details>
  <summary>Attempt to Delete Own Session</summary>

  <img src="https://github.com/user-attachments/assets/2815d6fe-05a7-4895-a0e9-b5238cf87702" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-15 at 23:23:15" />
</details>

<details>
  <summary>Delete 1 Session</summary>

  <img src="https://github.com/user-attachments/assets/be398820-6106-45da-b541-ab5af856be6b" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-15 at 23:23:19" />
</details>

<details>
  <summary>Delete All Sessions</summary>

  <img src="https://github.com/user-attachments/assets/bebfd1c2-cde7-4587-b29a-788897350ef6" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-15 at 23:23:29" />
</details>

<details>
  <summary>Showing off the Device Icons in Action</summary>

  <img src="https://github.com/user-attachments/assets/d337fa1e-e721-49ba-ae83-14bba0c0e8f0" alt="Simulator Screenshot - iPhone 16 Pro - 2024-10-15 at 23:24:51" />
  
</details>

LePips reviewed (changes requested):
Just an initial skim.

JPKribs commented on Shared/ViewModels/DevicesViewModel.swift:
```diff
@@ -0,0 +1,233 @@
+        let deviceIdsToDelete = self.devices.keys.filter { $0 != userSession.client.configuration.deviceID }
+
+        for deviceId in deviceIdsToDelete {
+            try await deleteDevice(id: deviceId)
```
I can switch that out!

What do you think of replacing Delete All with a multi-select? Since we're just passing a list into a single delete endpoint, I was thinking like what email has:

![image](https://github.com/user-attachments/assets/fae3ef50-57b8-4222-967f-43aecca9231f)

Then someone can select all > delete if they want to OR just delete specific ones instead of doing swipe, confirm, swipe, confirm... over and over.

JPKribs replied:
Hm... It might be too crowded to do something like this:
<img width="346" alt="Screenshot 2024-10-16 at 10 11 38" src="https://github.com/user-attachments/assets/24a1cec7-8fc3-4608-9ab1-4fcb49d53162">

I've updated all the ViewModel logic so the deleteDevices takes an array of ids then, DeleteAll just passes the self.devices in. So, if we want to go this route it's an easy functional change but UI-wise I'm struggling to get something together

LePips replied:
I agree that sounds like a good idea and good for at least having the initial functionality. I will also have to take a look at the button layouts.

JPKribs replied:
Sounds good. I fully got rid of Delete all in favor of:

`viewModel.send(.deleteDevices(ids: Array(viewModel.devices.keys)))`

The result is the same but it should make switching over to a multi-select really straight forward from a ViewModel perspective.

I wasn't sure if deleteDevice and deleteDevices was really necessary to have both. They're both in this most recent commit and it's all working as expected.

JPKribs replied:
This version works but the animations when going into editMode are a little jarring.

Having the **Select All** / **Remove All** at the top of the List is also not the easiest UX. Unfortunately, there just isn't enough toolbar room to fit everything.

<img width="1005" alt="Screenshot 2024-10-16 at 15 29 07" src="https://github.com/user-attachments/assets/b852025e-380e-4bb2-93d5-e57b6b3e5c66">

JPKribs replied:
This is what I have now:

<img width="350" alt="Screenshot 2024-10-16 at 18 55 35" src="https://github.com/user-attachments/assets/88974462-8045-4dbb-b8c0-5185ae7037e7">

<img width="344" alt="Screenshot 2024-10-16 at 18 57 18" src="https://github.com/user-attachments/assets/1d5a9b8f-91c2-4337-8109-5a2b97cf5a65">

<img width="343" alt="Screenshot 2024-10-16 at 18 57 25" src="https://github.com/user-attachments/assets/4356953b-5311-4e81-a9ab-5d9fc4066a42">

JPKribs replied:
Delete Devices might be a little much...

LePips replied:
Sorry, if you didn't see my edit: I meant hide the back bar button, not the bar. We can have the large delete button overlaid on the bottom like the user selection screen.

JPKribs replied:
Apologies, I think the screenshot looked weird. Those are on the bar:

```
        contentView
            .navigationTitle(L10n.allDevices)
            .navigationBarTitleDisplayMode(.inline)
            .navigationBarBackButtonHidden(selectMode)
            .toolbar {
                ToolbarItem(placement: .navigationBarLeading) {
                    if selectMode {
                        navigationBarSelectView
                    }
                }
                ToolbarItem(placement: .principal) {
                    if selectMode {
                        navigationBarDeleteView
                    }
                }
                ToolbarItem(placement: .navigationBarTrailing) {
                    navigationBarEditView
                }
            }
```

I've committed my current working version. At least, I *think* we're on the same page with what I did haha

JPKribs replied:
> Sorry, if you didn't see my edit: I meant hide the back bar button, not the bar. We can have the large delete button overlaid on the bottom like the user selection screen.

I need to replace this delete button let me re-look at this again today and try making this closer to SelectUserView

JPKribs replied:
Okay, apologies on this one. I see what you meant about the delete button. I've reworked my deviceRow to almost be a complete copy of the userRow from that selectUserView. I think there is only one 'issue' that's purely cosmetic. The TextPair doesn't let me darken the text. I can recreate this for this struct only and allow me to darken it but I wanted to make sure with you it was worth this effort:

<img width="340" alt="Screenshot 2024-10-17 at 13 06 21" src="https://github.com/user-attachments/assets/572694cd-7911-422c-9330-e3f78d0620dd">

<img width="345" alt="Screenshot 2024-10-17 at 13 07 41" src="https://github.com/user-attachments/assets/7423f354-fb68-4db0-9f45-71036837c199">

Other than that, this should almost exactly mirror selectUserView:

<details>
<summary>Screenshots</summary>
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 44](https://github.com/user-attachments/assets/bc01933c-6f4a-4b3a-85d9-cee7d7172cd5)
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 46](https://github.com/user-attachments/assets/6a0fbd45-ec8e-4335-8872-a56808cc4eb8)
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 49](https://github.com/user-attachments/assets/c6f74c20-3ff1-468d-be9e-8860e3f120e6)
![Simulator Screenshot - iPhone 16 Pro - 2024-10-17 at 13 07 53](https://github.com/user-attachments/assets/06d2464c-48bf-4360-ab31-212fba201d10)
<\details>

LePips replied:
`TextPairView` was using `foregroundColor`, not `foregroundStyle`. I've pushed those changes.

JPKribs commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/DevicesView.swift:
```diff
@@ -0,0 +1,201 @@
+import JellyfinAPI
+import SwiftUI
+
+struct DevicesView: View {
```
I was originally going to do a DeviceDetailView but I was finding that the only fields that weren't in use from the DeviceView we're very worthwhile:

<img width="463" alt="Screenshot 2024-10-16 at 08 54 05" src="https://github.com/user-attachments/assets/6dfa33fd-f841-4125-a73f-0156fb62555d">

I'd still be happy building out a DeviceDetailView but I think it'd be the same information + once we the SDK updates to include customName (its on main now) I can add that as well. 

Your call!

JPKribs replied:
> We could also use the last used user id and show basic user information.

I'm adding some changes to this commit that are for a later PR:

One of the other items I have built out in another branch is the Users section. That would show all Jellyfin Users. 

The UserDetails has Forms for changing Password & Permissions, and a listView for Devices. That calls this same view but filters on UserID.

So I was thinking instead of going Devices > Users we could go from Users > Devices then have this same view reused there to allow changes to specific user devices.

LePips replied:
Even if it's just a few more fields, I still prefer it over the alert viewing for changing the name.

Per the user comment, I'm not saying we should push to `Users` from this view. We could just get the single-user data and just show their profile picture and name like we do in sessions. Then, we should push to the user detail view from there.

JPKribs replied:
Sounds good! Does this look like everything we want?

![Simulator Screenshot - iPhone 16 Pro - 2024-10-16 at 12 27 22](https://github.com/user-attachments/assets/2f3f906e-e61e-4d08-ab64-6e0457587951)

LePips replied:
That looks great!

JPKribs commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/DevicesView.swift:
```diff
@@ -0,0 +1,201 @@
+    // MARK: - Content View
+
+    private var contentView: some View {
+        Group {
```
Ignore that. @ViewBuilder...

JPKribs commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/DevicesView.swift:
```diff
@@ -0,0 +1,201 @@
+                L10n.devices,
+                description: L10n.allDevicesDescription
+            ) {
+                UIApplication.shared.open(URL(string: "https://jellyfin.org/docs/general/server/devices")!)
```
I did this for this PR plus the one for scheduled tasks as well. I've got a couple other things I'd like to be get done in the admin dashboard and I will make sure I do the same for that as well.

JPKribs commented:
I've cleaned this all up from my end and added the DeviceDetailsView. I left a TODO at:

`Swiftfin/Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/Components/Sections/CustomDeviceNameSection.swift`

`// TODO: Enable for CustomNames for Devices: 27, 50, & 53`

The version of the SDK we are on does not have CustomName available from DeviceInfo but it IS available on the Main version of the SDK. This TODO is just to uncomment out those lines and this should be good to go.

Lastly, I added DeviceCompatibilities to the view since it was looking a little bare. I don't think it's anything revolutionary but some people might find it valuable...

I split this out into Sections so, down the road, if we do have activity logs that we can join to Device/UserId, it should be pretty easy to slot that in here.

LePips commented on Shared/ViewModels/DevicesViewModel.swift:
```diff
@@ -0,0 +1,251 @@
+    enum Action: Equatable {
+        case getDevices(_ userID: String?)
+        case setCustomName(id: String, newName: String)
+        case deleteDevice(id: String)
```
Don't need the single id instance.

LePips commented on Shared/ViewModels/DevicesViewModel.swift:
```diff
@@ -0,0 +1,251 @@
+    // MARK: - Action
+
+    enum Action: Equatable {
+        case getDevices(_ userID: String?)
```
Instead, this class should take the user ID in its `init` so that it's always associated with it.

LePips commented on Shared/ViewModels/DevicesViewModel.swift:
```diff
@@ -0,0 +1,251 @@
+
+    // MARK: - Delete Device
+
+    private func deleteDevice(id: String) async throws {
```
Don't need single id instance.

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/Components/Sections/CompatibilitiesSection.swift:
```diff
@@ -0,0 +1,36 @@
+        var device: DeviceInfo
+
+        var body: some View {
+            Section(header: Text(L10n.capabilities)) {
```
```suggestion
            Section(L10n.capabilities) {
```

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/Components/Sections/CustomDeviceNameSection.swift:
```diff
@@ -0,0 +1,58 @@
+    struct CustomDeviceNameSection: View {
+        var device: DeviceInfo
+
+        private var viewModel = DevicesViewModel()
```
Should be a `StateObject`.

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/Components/Sections/CustomDeviceNameSection.swift:
```diff
@@ -0,0 +1,58 @@
+
+        var body: some View {
+            Section(L10n.customDeviceName) {
+                ChevronAlertButton(
```
Instead, we should have a `TextField` as a row and have an external save button, like user pin hint/password. Have the save button in the top right corner (user password should be migrated to that).

JPKribs replied:
I was stealing some of the eventful configuration from the resetUserPasswordViewModel. I'm incorporated that here as well:

https://github.com/user-attachments/assets/4daa5726-0d0f-4a11-a6ca-43c169d21ffd

JPKribs replied:
I also added this disclaimer just to try and limit any future issues that might get generated from this. I think after this PR I might make an issue in advance just to get ahead of it:

<img width="326" alt="Screenshot 2024-10-16 at 17 37 43" src="https://github.com/user-attachments/assets/8b1f64a0-8b04-4cce-95dd-ceb2d64a4c04">

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/Components/Sections/UserSection.swift:
```diff
@@ -0,0 +1,36 @@
+        var device: DeviceInfo
+
+        var body: some View {
+            Section(header: Text(L10n.user)) {
```
```suggestion
            Section(header: L10n.user) {
```

LePips reviewed (changes requested):
Another look, I'll take a closer look at all strings later.

LePips commented on Shared/Strings/Strings.swift:
```diff
@@ -32,6 +32,10 @@ internal enum L10n {
     return L10n.tr("Localizable", "airWithDate", p1, fallback: "Airs %s")
   }
+  /// All Devices
+  internal static let allDevices = L10n.tr("Localizable", "allDevices", fallback: "All Devices")
```
Just use the `Devices` string instead.

LePips commented on Shared/Strings/Strings.swift:
```diff
@@ -276,6 +304,8 @@ internal enum L10n {
   /// Enabled
   internal static let enabled = L10n.tr("Localizable", "enabled", fallback: "Enabled")
+  /// Supply a custom display name or leave empty to use the name reported by the device.
+  internal static let enterCustomDeviceName = L10n.tr("Localizable", "enterCustomDeviceName", fallback: "Supply a custom display name or leave empty to use the name reported by the device.")
```
I want to reduce "leave x in a certain value". Instead, let's have a toggle for using a custom name and we will update the name accordingly.

LePips commented on Shared/ViewModels/DevicesViewModel.swift:
```diff
@@ -0,0 +1,255 @@
+
+    // MARK: - Initializer
+
+    init(_ userID: String?) {
```
```suggestion
    init(_ userID: String? = nil) {
```

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/Components/Sections/CustomDeviceNameSection.swift:
```diff
@@ -0,0 +1,35 @@
+                )
+            } header: {
+                Text(L10n.customDeviceName)
+                // TODO: Remove Footer after SDK Changes
```
If this just doesn't work, then I don't want to have it surfaced in the app. We can just comment out the place where this is used and have a todo to uncomment when we update the SDK.

JPKribs replied:
The SDK lets us update the customName and the customName shows up on Web and other places where this is used. But, our current version of the SDK doesn't have customName included. So we can **set** it but we can't **see** it. 

That'd be commenting out the edit button and and the textField. Do you want me to proceed?

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/Components/Sections/UserSection.swift:
```diff
@@ -0,0 +1,36 @@
+
+extension DeviceDetailsView {
+    struct UserSection: View {
+        var device: DeviceInfo
```
```suggestion
        let device: DeviceInfo
```

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/DeviceDetailsView.swift:
```diff
@@ -0,0 +1,99 @@
+
+struct DeviceDetailsView: View {
+
+    var device: DeviceInfo
```
```suggestion
    let device: DeviceInfo
```

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DeviceDetailsView/DeviceDetailsView.swift:
```diff
@@ -0,0 +1,99 @@
+            }
+        }
+        .topBarTrailing {
+            HStack {
```
Doesn't need to be in an `HStack`

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/Components/DeviceRow.swift:
```diff
@@ -0,0 +1,198 @@
+        // MARK: - Device View
+
+        @ViewBuilder
+        private var deviceView: some View {
```
Not used.

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/Components/DeviceRow.swift:
```diff
@@ -0,0 +1,198 @@
+                }
+                .tint(.red)
+            }
+            .contextMenu {
```
We don't need both swipe action and context menu, just have the swipe action delete.

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/Components/DeviceRow.swift:
```diff
@@ -0,0 +1,198 @@
+            }
+            .onSelect(perform: onSelect)
+            .swipeActions {
+                Button {
```
`Button(L10n.delete, systemName: "trash", action: onDelete)`

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/DevicesView.swift:
```diff
@@ -0,0 +1,245 @@
+    @State
+    private var selectedDevices: Set<String> = []
+    @State
+    private var editMode: Bool = false
```
```suggestion
    private var isEditing: Bool = false
```

Although, I should refactor to use the environment `editMode`.

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/DevicesView.swift:
```diff
@@ -0,0 +1,245 @@
+
+    @ViewBuilder
+    private var navigationBarSelectView: some View {
+        Button(selectedDevices == Set(viewModel.devices.keys) ? L10n.removeAll : L10n.selectAll) {
```
Instead, condition on the `count` of each instead of making the set.

LePips commented on Swiftfin/Views/SettingsView/UserDashboardView/DevicesView/DevicesView.swift:
```diff
@@ -0,0 +1,245 @@
+
+    // MARK: - Deletion Failure Alert
+
+    private var deletionFailureAlert: Alert {
```
The `Alert` object is deprecated, we shouldn't be making one.

JPKribs commented:
I have these review changes merged in. I believe this should be all good! Thank you for your help and patience @LePips. I feel like this addition ended up in a really good place!

LePips commented:
I came to perform the final review/finishing touches and came into a design conundrum, realizing that the `Active Devices` and `Devices` have different list styles, which I didn't like. We want the header information which we achieve with `List` but we use `CollectionVGrid` for `Active Devices`. We aren't able to introspect the layout to achieve the inset grouped for the header and the plain list style for the rows, which means we will have to roll our own layout.

I was able to make a `UICollectionViewCompositionalLayout` that should be able to achieve what we want. I will take a look at incorporating this into my `CollectionVGrid` package and have this view use it here, and we may want to change `Active Devices` to use that as well. I may also look at renaming that action to `Sessions`.

JPKribs commented:
I can make the name change and flip this to use the collection V grid (for now without the header). I just wanted to make sure you didn't have changes on a branch before I started making changes on my end and screwing that up that commit?

Lemme know and I can swap these around if needed!

LePips commented:
I can do the rest of the work necessary for this. The renaming from `Active Devices` to `Sessions` can be done and approved separately.

JPKribs commented:
> I can do the rest of the work necessary for this.

Sounds good. I appreciate your help getting this (and many other PRs) over the finish line!

LePips commented:
I did implement what I said I would implement: https://github.com/LePips/CollectionVGrid/commit/65fbe9be0f73fc8fa780d40be3d8710c5be115fe, however there are issues when performing the editing that would require some more significant refactoring.

Instead, I just made the list header look like it's inset grouped but the colors are inverted in light mode when compared to the normal inset grouped style, which is okay. We will still use `List` and luckily these are still all list-styled, however this may change on iPadOS if we want to change the design to be full-screen and have a grid layout. That's an issue for later.
